src/python/TaskWorker/Actions/Recurring/TapeRecallManager.py

```diff
@@ -0,0 +1,239 @@
+    name of this class, this file and the recurring action in TW config list
+    must be the same """
+    pollingTime = 60*4  # unit=minutes. Runs every 4 hours
+    rucioClient = None
+    privilegedRucioClient = None
+    crabserver = None
+
+    def _execute(self, config, task):  # pylint: disable=unused-argument
+        """ this is what we do at every polling cycle """
+        self.config = config
+        # setup logger, crabserver client, rucio client
+        self.init()
+        # do the work
+        self.handleRecall()
+        self.handleKill()
+
+
+    def handleKill(self):
+        """ looks for tasks in KILLRECALL and deals with them """
+        status = 'KILLRECALL'
+        tasksToWorkOn = self.getTasks(status)
+
+        for aTask in tasksToWorkOn:
+            taskName = aTask['tm_taskname']
+            self.logger.info("Working on task %s", taskName)
+
+            reqId = aTask['tm_DDM_reqid']
+            if not reqId:
+                self.logger.debug("tm_DDM_reqid' is not defined for task %s, skipping it", taskName)
+                continue
+            msg = "Task points to Rucio RuleId:  %s " % reqId
+            self.logger.info(msg)
+
+            # delete rule and set task status to killed
+            self.logger.info('Will delete rule %s', reqId)
+            self.privilegedRucioClient.delete_replication_rule(reqId)
+            self.updateTaskStatus(taskName, 'KILLED')
+            # Clean up previous "dataset on tape" warnings
+            self.deleteWarnings(taskName)
+            self.logger.info("Done on this task")
+
+
+    def handleRecall(self):
+        """ looks for tasks in TAPERECALL and deals with them """
+        status = 'TAPERECALL'
+        tasksToWorkOn = self.getTasks(status)
+        for aTask in tasksToWorkOn:
+            taskName = aTask['tm_taskname']
+            self.logger.info("Working on task %s", taskName)
+            # 1.) check for "waited too long"
+            if (time.time() - getTimeFromTaskname(str(taskName))) > MAX_DAYS_FOR_TAPERECALL*24*60*60:
+                msg = "Disk replica request (ID: %s) for input data did not complete in %d days." %\
+                      (reqId, MAX_DAYS_FOR_TAPERECALL)
+                self.logger.info(msg)
+                failTask(taskName, self.crabserver, msg, self.logger, 'FAILED')
+                continue
+            # 2.) integrity checks
+            reqId = aTask['tm_DDM_reqid']
+            if not reqId:
+                self.logger.debug("tm_DDM_reqid' is not defined for task %s, skipping it", taskName)
+                continue
+            self.logger.info("Task points to Rucio RuleId:  %s ", reqId)
+            try:
+                rule = self.rucioClient.get_replication_rule(reqId)
+            except RuleNotFound:
+                msg = "Rucio rule id %s not found. Please report to experts" % reqId
+                self.logger.error(msg)
+                self.uploadWarning(msg, taskName)
+                continue
```

Comment:
One case that can get weird, but not bug, is when `MAX_DAYS_FOR_TAPERECALL` in condition line 79 is *more than* rule expiration time due to misconfiguration. Worst case is TaskWorker keep upload same error every 4 hour.
I am not sure if we need to take any action or add a bit of comment here or it can silently ignore.

Reply:
I agree that it is a misconfiguration, then it is OK to raise an error so that someone reports it and we can fix it. We do not have a machinery in place to have such problems automatically reported to CRAB DevOps !  CRAB Server has a limit to how many messages it accepts so this will not flood DB nor user's console.
That's a general/generic line of thought. On this I think we can code so that it can't happen. Let me look at it.